ui/webui/test/run-tests

```diff
@@ -0,0 +1,470 @@
+    retry_reason = ""
+    skip_reason = ""
+
+    # Try affected tests 3 times
+    if test.process.returncode == 0 and affected and test.retry_when_affected and test.retries < 2:
+        retry_reason = "test affected tests 3 times"
+        test.retries += 1
+        print_test(test, not opts.list, "# RETRY {0} ({1})".format(test.retries, retry_reason))
+        return retry_reason, 0
+
+    # If test is being skipped pick up the reason
+    if test.process.returncode == 77:
+        lines = test.output.splitlines()
+        skip_reason = lines[-1].strip().decode("utf-8")
+        test.output = b"\n".join(lines[:-1])
+
+    if test.process.returncode in [0, 77]:
+        print_test(test, not opts.list, skip_reason=skip_reason)
+        return None, 0
+
+    if not opts.thorough:
+        cmd = ["test-failure-policy", testvm.DEFAULT_IMAGE]
+        try:
+            proc = subprocess.Popen(cmd, stdin=subprocess.PIPE, stdout=subprocess.PIPE)
+            reason = proc.communicate(test.output + ("not ok " + test_name(test)).encode())[0].strip()
+
+            if proc.returncode == 77:
+                print_test(test, skip_reason="# SKIP {0}".format(reason.decode("utf-8")))
+                return None, 0
+
+            if proc.returncode == 1:
+                retry_reason = reason.decode("utf-8")
+
+        except OSError as ex:
+            if ex.errno != errno.ENOENT:
+                sys.stderr.write("\nCouldn't run test-failure-policy: {0}\n".format(str(ex)))
+
+    # HACK: many tests are unstable, always retry them 3 times unless affected
+    if not affected and not retry_reason:
+        retry_reason = "be robust against unstable tests"
+
+    unexpected_message = testlib.UNEXPECTED_MESSAGE.encode() in test.output
+    if test.retries < 2 and not unexpected_message and retry_reason:
+        test.retries += 1
+        print_test(test, retry_reason="# RETRY {0} ({1})".format(test.retries, retry_reason))
+        return retry_reason, 0
+
+    test.output += b"\n"
+    print_test(test)
+    return None, 1
+
+
+def check_valid(filename):
+    name = os.path.basename(filename)
+    allowed = string.ascii_letters + string.digits + '-_'
+    if not all(c in allowed for c in name):
+        return None
+    return name.replace("-", "_")
+
+
+def build_command(filename, test, opts):
+    cmd = [filename]
+    if opts.trace:
+        cmd.append("-t")
+    if opts.verbosity:
+        cmd.append("-v")
+    if not opts.fetch:
+        cmd.append("--nonet")
+    if opts.list:
+        cmd.append("-l")
+    cmd.append(test)
+    return cmd
+
+
+class GlobalMachine:
+    def __init__(self):
+        self.image = testvm.DEFAULT_IMAGE
+        self.machine = machine_install.VirtInstallMachine(image=self.image)
+        if not os.path.exists(self.machine.image_file):
+            self.machine.pull(self.machine.image_file)
+        self.machine.start()
+
+    def kill(self):
+        self.machine.kill()
+
+
+def run(opts, image):
```

Comment:
I would love to have this split into smaller functions. It's too big now.

Reply:
This file as I mention in the start, is a copy paste from cockpit, with some sligh adjustments. The original script needs to be adjusted and this is about to be deleted soon.

Reply:
Ahh, ok thanks for explanation.